src/Native/Unix/System.Native/pal_console.c

```diff
@@ -289,57 +320,76 @@ int32_t SystemNative_SetSignalForBreak(int32_t signalForBreak)
     {
-        IncorporateBreak(&current, signalForBreak);
-        if (tcsetattr(STDIN_FILENO, TCSANOW, &current) >= 0)
+        struct termios termios;
+        if (TcGetAttr(&termios, signalForBreak, g_hasInteractiveChildren))
         {
-            g_signalForBreak = signalForBreak;
-            return 1;
+            if (TcSetAttr(&termios))
+            {
+                g_signalForBreak = signalForBreak;
+                rv = 1;
+            }
         }
+        pthread_mutex_unlock(&g_lock);
     }
 
-    return 0;
+    return rv;
 }
 
 void ReinitializeConsole()
 {
-    // pal_signal.cpp calls this on SIGCONT/SIGCHLD.
-    // This can happen when SystemNative_InitializeConsole was not called.
-    // This gets called on a signal handler, we may only use async-signal-safe functions.
-
-    // If the process was suspended while reading, we need to
-    // re-initialize the console for the read, as the attributes
-    // previously set were likely overwritten.
-    if (g_readInProgress)
+    // Restores the state of the console after being suspended.
+    // pal_signal.cpp calls this on SIGCONT from the signal handling thread.
+
+    if (pthread_mutex_lock(&g_lock) == 0)
     {
-        IncorporateBreak(&g_currTermios, g_signalForBreak);
-        tcsetattr(STDIN_FILENO, TCSANOW, &g_currTermios);
-    }
+        struct termios termios;
+        if (TcGetAttr(&termios, g_signalForBreak, g_hasInteractiveChildren))
+        {
+            TcSetAttr(&termios);
+        }
 
-    // "Application mode" will also have been reset and needs to be redone.
-    WriteKeypadXmit();
+        // Redo "Application mode" unless there are interactive children.
+        // In that case, we'll redo it when there are no more interactive children.
+        if (!g_hasInteractiveChildren)
+        {
+            WriteKeypadXmit();
+        }
+
+        pthread_mutex_unlock(&g_lock);
+    }
 }
 
-int32_t SystemNative_InitializeConsole()
+int32_t SystemNative_InitializeConsoleAndSignalHandling()
 {
-    if (!InitializeSignalHandling())
-    {
-        return 0;
-    }
+    static int32_t initialized = 0;
 
-    if (tcgetattr(STDIN_FILENO, &g_initTermios) >= 0)
-    {
-        g_haveInitTermios = true;
-        g_signalForBreak = (g_initTermios.c_lflag & ISIG) != 0;
-    }
-    else
+    // Both the Process and Console class call this method for initialization.
+    if (pthread_mutex_lock(&g_lock) == 0)
     {
-        g_haveInitTermios = false;
-        g_signalForBreak = true;
+        if (initialized == 0)
+        {
+            initialized = InitializeSignalHandlingCore();
+
+            if (initialized == 1)
+            {
+                g_haveInitTermios = tcgetattr(STDIN_FILENO, &g_initTermios) >= 0;
+
+                if (g_haveInitTermios)
+                {
+                    struct termios termios = g_initTermios;
+                    ApplyTerminalSettings(&termios, g_signalForBreak, g_hasInteractiveChildren);
```

Comment:
I've made a functional change here.
Before on initialization, we'd set `g_signalForBreak` (`Console.TreatControlCAsInput`) based on the terminal settings. Now we explicitly enforce a value.